File: frontend/src/lib/callTrail.ts
// THE CALL TRAIL, browser half (D77) — a buffered, debug-only record of what the call machine decided.
//
// Every decision that matters in a live call is made HERE, in the browser — the relative gate's floor,
// the noise and voice estimates, the leak probe's verdicts, the ear hold, the transcript gate's drops,
// the route readbacks, the phase changes — and the relay sees only the wire. So `useLiveCall` pushes
// what it decided into this buffer, and the buffer ships it to `POST /api/voice/live/trail`, which
// appends it to the SAME per-call file the relay writes (`$CTRLB_HOME/calls/<callId>.jsonl`). The main
// seat reads the file after the owner's phone round instead of reconstructing the call from memory.
//
// PURE-ISH ON PURPOSE: no React, no store, no fetch of its own — the POST is injected (`post`), the
// clock is injectable (`now`), and the ONE timer is the flush interval. Nothing renders off it.
//
// DEBUG DATA, SHAPED LIKE IT. A rejected POST drops its batch (one `console.debug`, never a retry, never
// a throw into the call); a disposed trail ignores every push. The call must never notice the trail.

/** How often the buffer flushes on its own, ms. A property of READING the trail afterwards, not of the
 *  call: two seconds of lines is fine-grained enough to line up with the relay's own stamps and coarse
 *  enough that a call makes one small POST at a time. Started lazily by the first push, cleared when a
 *  flush empties the buffer — an idle call makes no requests at all. */
export const TRAIL_INTERVAL_MS = 2000;

/** Entries per batch before the buffer flushes early (`"count"`). */
export const TRAIL_MAX_ENTRIES = 50;

/** SERIALIZED bytes per batch before the buffer flushes early (`"count"`, council F3). The browser's
 *  `keepalive` budget is ~64 KiB per ORIGIN across every in-flight keepalive request, and the `hidden`
 *  / `end` flushes are exactly the ones that ride it — so one batch stays at half of it. The server's
 *  64 KB body cap is the hard bound beside this soft one. */
export const TRAIL_FLUSH_BYTES = 32 * 1024;

/** Entries per POST — the trail route's own `entries` bound. A flush larger than this splits. */
export const TRAIL_POST_ENTRIES = 200;

/** SERIALIZED size of ONE entry — the trail route's own per-entry bound (`TRAIL_MAX_ENTRY_BYTES` in
 *  `services/call_trail.py`), mirrored like `TRAIL_POST_ENTRIES`. The server refuses the WHOLE batch
 *  over one entry past it, and a line can copy an unbounded upstream text (a `serverError`'s message),
 *  so an entry past it is replaced by a marker at push — see `push`. */
export const TRAIL_MAX_ENTRY_BYTES = 2048;

export type TrailFlushReason = "interval" | "count" | "hidden" | "end";

export interface CallTrail {
  /** Buffer one line: `{t, leg, gen, ev, ...data}` — stamped at push time. */
  push(ev: string, data?: Record<string, unknown>): void;
  /** Ship the buffer now. `hidden`/`end` ride `keepalive`, so the POST outlives a page going away. */
  flush(reason: TrailFlushReason): void;
  /** Stop the timer and ignore every later push. Idempotent. */
  dispose(): void;
}

export interface CallTrailOpts {
  callId: string;
  /** The injected write — `postJSON(<route>, body, { keepalive })` in production. */
  post: (body: unknown, keepalive: boolean) => Promise<void>;
  /** WHICH leg and generation a line belongs to, read at PUSH time (council F4): one call is many legs
   *  (reconnects, route cycles), and the relay's lines carry the same `leg`. */
  stamp: () => { leg: number; gen: number };
  now?: () => number;
  intervalMs?: number;
  maxEntries?: number;
}

export function createCallTrail(opts: CallTrailOpts): CallTrail {
  const now = opts.now ?? Date.now;
  const intervalMs = opts.intervalMs ?? TRAIL_INTERVAL_MS;
  const maxEntries = opts.maxEntries ?? TRAIL_MAX_ENTRIES;
  let buf: Record<string, unknown>[] = [];
  /** The serialized size of `buf` so far — each entry measured ONCE, at push. */
  let bytes = 0;
  let timer: ReturnType<typeof setInterval> | undefined;
  let disposed = false;
  let warned = false;

  const stopTimer = (): void => {
    clearInterval(timer);
    timer = undefined;
  };

  const drop = (e: unknown): void => {
    if (warned) return;
    warned = true;
    console.debug("call trail: a batch was dropped (debug data, not retried)", e);
  };

  const ship = (entries: Record<string, unknown>[], keepalive: boolean): void => {
    for (let i = 0; i < entries.length; i += TRAIL_POST_ENTRIES) {
      const body = { call_id: opts.callId, entries: entries.slice(i, i + TRAIL_POST_ENTRIES) };
      // `.catch` on the promise AND a try around the call: a `post` that throws synchronously must
      // not reach the call machine either.
      try {
        void opts.post(body, keepalive).catch(drop);
      } catch (e) {
        drop(e);
      }
    }
  };

  const flush = (reason: TrailFlushReason): void => {
    stopTimer();
    if (buf.length === 0) return;
    const entries = buf;
    buf = [];
    bytes = 0;
    ship(entries, reason === "hidden" || reason === "end");
  };

  return {
    push(ev, data) {
      if (disposed) return;
      const head = { t: now(), ...opts.stamp(), ev };
      let entry: Record<string, unknown> = { ...head, ...data };
      let json = JSON.stringify(entry).length;
      // ONE OVERSIZED LINE MUST NOT COST ITS NEIGHBOURS: the server 422s a batch over any entry past
      // its per-entry bound, which would drop every other line of that batch with it. So the entry is
      // replaced by a marker that still says WHAT happened and WHEN (and how big it was). `.length`
      // counts UTF-16 units, never fewer than the characters the server counts — the check errs early.
      if (json > TRAIL_MAX_ENTRY_BYTES) {
        // …and WHICH signal it was, when the line is a `sig`: the type is the one field that names it.
        const type = data?.type;
        entry = { ...head, ...(typeof type === "string" ? { type } : {}), oversize: json };
        json = JSON.stringify(entry).length;
      }
      const size = json + 1; // + the array comma
      // BYTES FIRST: the entry that would push the batch past the keepalive half goes into the NEXT
      // batch, so no batch ever exceeds it (every entry is within the per-entry bound above, far
      // inside the half).
      if (buf.length > 0 && bytes + size > TRAIL_FLUSH_BYTES) flush("count");
      buf.push(entry);
      bytes += size;
      if (buf.length >= maxEntries) {
        flush("count");
        return;
      }
      if (timer === undefined) timer = setInterval(() => flush("interval"), intervalMs);
    },
    flush(reason) {
      if (disposed) return;
      flush(reason);
    },
    dispose() {
      disposed = true;
      stopTimer();
      buf = [];
      bytes = 0;
    },
  };
}

File: backend/app/services/call_trail.py
"""The CALL TRAIL (D77) — an append-only, per-call JSONL record of a live call, for diagnosis.

Every decision that matters in a live call is made in the BROWSER (the relative gate's floor, the leak
probe's verdicts, the ear hold, the transcript gate's drops, the route readbacks) and the relay sees
only the wire — so until this module there was no record of a call anywhere, and the owner narrated
phone sessions from memory. The trail is written by BOTH halves into one file per call, keyed by the
`call_id` the client mints: the relay's lines (`src: "relay"`, `services/voice_live.py`) and the
browser's (`src: "client"`, through `POST /api/voice/live/trail`). The main seat reads
`$CTRLB_HOME/calls/<call_id>.jsonl` after the fact.

**Debug data, and shaped like it.** Gated by `voice.live.debug` (off by default — the same knob that
turns the on-screen readout on); no fsync, because a lost tail on a power cut is a lost diagnostic,
not lost data; and a write failure (disk full, a read-only mount) is logged ONCE and swallowed —
a trail must never end a call. Retention is by count (`voice.live.trail_keep`), pruned when a NEW
call's first line lands, so the directory cannot grow without bound however many calls are made.

**The filename is the path-traversal guard.** `call_id` comes off the wire (the relay's `start`
control, the trail route's body), so it is validated against ONE canonical-UUID pattern before it
becomes a path — defined here, imported by both callers, never re-spelled.

Synchronous on purpose (tiny appends): both async callers go through `asyncio.to_thread`, so the
relay's audio pump never waits on a disk. The lock is what lets the two writers — the relay's thread
and the route's — share one file without interleaving inside a line.
"""

from __future__ import annotations

import json
import logging
import os
import re
import threading
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Iterable, Mapping
    from pathlib import Path

log = logging.getLogger(__name__)

#: A canonical lowercase UUID — what `crypto.randomUUID()` mints. ONE spelling for the three places
#: that validate a call id: this store (the path guard), the relay's `start` parse and the trail
#: route's body model (pydantic takes the string; Python callers use `CALL_ID_RE.fullmatch`, never
#: `match`, because `$` would admit a trailing newline).
CALL_ID_PATTERN = r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
CALL_ID_RE = re.compile(CALL_ID_PATTERN)

#: The trail route's wire bounds (D77). Not config: they bound what one debug POST may cost, and the
#: client's own flush budget (32 KiB — the browser's keepalive quota is ~64 KiB per origin across
#: in-flight requests) sits well inside them. A whole body past `TRAIL_MAX_BODY_BYTES` is a 413; one
#: entry past `TRAIL_MAX_ENTRY_BYTES` (serialized) is a 422 naming its index; a batch is at most
#: `TRAIL_MAX_ENTRIES` lines.
TRAIL_MAX_BODY_BYTES = 64 * 1024
TRAIL_MAX_ENTRY_BYTES = 2048
TRAIL_MAX_ENTRIES = 200


def valid_call_id(value: object) -> bool:
    """Is `value` a call id this store will turn into a filename?"""
    return isinstance(value, str) and CALL_ID_RE.fullmatch(value) is not None


class CallTrail:
    """Append-only per-call JSONL under `$CTRLB_HOME/calls/` — debug data, see D77.

    Holds no settings: `keep` is passed to every `append` (read live from `voice.live.trail_keep` by
    the caller, the `LiveSessionSlots` precedent), so one instance lives for the process and a Conf
    edit needs no rebuild.
    """

    def __init__(self, root: Path) -> None:
        self._root = root
        self._lock = threading.Lock()
        #: ONE warning per process for a failing disk — a trail that cannot write says so once, not
        #: once per batch for the rest of the call.
        self._warned = False

    @property
    def root(self) -> Path:
        return self._root

    def append(self, call_id: str, lines: Iterable[Mapping[str, Any]], *, keep: int) -> None:
        """Append `lines` to `<root>/<call_id>.jsonl`, one compact JSON object per line.

        Raises `ValueError` for a malformed `call_id` (a caller bug or a hostile id — never a path).
        Every I/O failure is swallowed after the one warning. When this append CREATED the file, the
        directory is pruned to the newest `keep` trails by mtime — a second append to the same call
        never prunes, so a long call cannot delete its own history mid-flight.
        """
        if not valid_call_id(call_id):
            raise ValueError("call_id must be a canonical lowercase UUID")
        payload = "".join(
            json.dumps(line, separators=(",", ":"), ensure_ascii=False) + "\n" for line in lines
        ).encode("utf-8")
        if not payload:
            return
        path = self._root / f"{call_id}.jsonl"
        with self._lock:
            try:
                self._root.mkdir(mode=0o700, parents=True, exist_ok=True)
                created = not path.exists()
                # 0o600: the lines carry the owner's transcripts. O_APPEND, one batch per open; the
                # buffered writer loops a short `write(2)` to completion or raises (the S3 code round,
                # F3 — a bare `os.write` may return early and would truncate a line in silence).
                fd = os.open(path, os.O_WRONLY | os.O_APPEND | os.O_CREAT, 0o600)
                with os.fdopen(fd, "ab") as f:
                    f.write(payload)
                if created:
                    self._prune(keep, current=path)
            except OSError:
                if not self._warned:
                    self._warned = True
                    log.warning(
                        "call trail: cannot write under %s (further failures are silent)",
                        self._root,
                        exc_info=True,
                    )

    def _prune(self, keep: int, *, current: Path) -> None:
        """Keep `current` plus the newest `keep - 1` OTHER trails by mtime; unlink the rest (best-effort,
        per file). `current` is never a candidate (the S3 code round, F5): on a coarse-mtime filesystem
        the file just created can tie with an older one, and a tie must not delete the live call."""
        others = sorted((p for p in self._root.glob("*.jsonl") if p != current), key=_mtime, reverse=True)
        for stale in others[max(keep - 1, 0) :]:
            try:
                stale.unlink()
            except OSError:
                log.debug("call trail: could not prune %s", stale.name, exc_info=True)


def _mtime(path: Path) -> float:
    try:
        return path.stat().st_mtime
    except OSError:  # vanished between the glob and the stat — sorts last, then unlink misses it
        return 0.0
